src/main/java/org/openrewrite/java/testing/cleanup/RemoveVisibleForTestingAnnotationWhenUsedInProduction.java

```diff
@@ -0,0 +1,196 @@
+            }
+
+            @Override
+            public J.NewClass visitNewClass(J.NewClass newClass, ExecutionContext ctx) {
+                // Mark constructors
+                if (newClass.getConstructorType() != null) {
+                    if(!acc.methods.contains(newClass.getConstructorType())) {
+                        newClass.getConstructorType().getAnnotations().forEach(annotation -> {
+                            if ("VisibleForTesting".equals(annotation.getClassName())) {
+                                J.CompilationUnit compilationUnit = getCursor().firstEnclosing(J.CompilationUnit.class);
+                                if (compilationUnit != null && compilationUnit.getSourcePath().startsWith("src/main")) {
+                                    acc.methods.add(newClass.getConstructorType());
+                                }
+                            }
+                        });
+                    }
+                }
+                // Mark classes
+                if (newClass.getClazz() != null && newClass.getClazz().getType() != null && newClass.getClazz().getType() instanceof JavaType.Class) {
+                    JavaType.Class clazzType = (JavaType.Class) newClass.getClazz().getType();
+                    if (!acc.classes.contains(clazzType)) {
+                        clazzType.getAnnotations().forEach(annotation -> {
+                            if ("VisibleForTesting".equals(annotation.getClassName())) {
+                                J.CompilationUnit compilationUnit = getCursor().firstEnclosing(J.CompilationUnit.class);
+                                if (compilationUnit != null && compilationUnit.getSourcePath().startsWith("src/main")) {
+                                    acc.classes.add(clazzType);
+                                }
+                            }
+                        });
+                    }
+                }
+                return super.visitNewClass(newClass, ctx);
+            }
+        };
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor(Scanned acc) {
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            @Override
+            public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations multiVariable, ExecutionContext ctx) {
+                J.VariableDeclarations variableDeclarations = super.visitVariableDeclarations(multiVariable, ctx);
+                if(!variableDeclarations.getVariables().isEmpty()) {
+                    // if none of the variables in the declaration are used from production code, the annotation should be kept
+                    boolean keepAnnotation = variableDeclarations.getVariables().stream().noneMatch(elem -> acc.fields.contains(elem.getVariableType()));
+                    if (!keepAnnotation) {
+                        Optional<J.Annotation> annotation = variableDeclarations.getLeadingAnnotations().stream()
```

Comment:
Lot's of repetitive code here. I think you go very far with a generic function. Just out the top of my head:

```java
private <@Nullable T extends J> getElement(List<J.Annotation> leadingAnnotations) {
    Optional<J.Annotation> annotation = leadingAnnotations().stream() .filter(elem -> "VisibleForTesting".equals(elem.getSimpleName())).findFirst();
    if (annotation.isPresent() && annotation.get().getType() instanceof JavaType.Class) {
        JavaType.Class type = (JavaType.Class) annotation.get().getType();
        return (T) new RemoveAnnotation("@" + type.getFullyQualifiedName()).getVisitor().visitNonNull(variableDeclarations, ctx, getCursor().getParentOrThrow());
    }

    return null;
}